internal/impl/mssqlserver/replication/snapshot.go

```diff
@@ -0,0 +1,341 @@
+	log       *service.Logger
+}
+
+// NewSnapshot creates a new instance of Snapshot capable of snapshotting provided tables.
+// It does this by creating a transaction with snapshot level isolation before paging through rows, sending them to be batched.
+func NewSnapshot(db *sql.DB, tables []UserTable, publisher ChangePublisher, logger *service.Logger) *Snapshot {
+	return &Snapshot{
+		db:        db,
+		Tables:    tables,
+		publisher: publisher,
+		log:       logger,
+	}
+}
+
+// Prepare performs initial validation, creating of connections in preparation for snapshotting tables.
+func (s *Snapshot) Prepare(ctx context.Context) (LSN, error) {
+	if len(s.Tables) == 0 {
+		return nil, errors.New("no tables provided")
+	}
+
+	var err error
+
+	// create separate connection for snapshotting
+	if s.snapshotConn, err = s.db.Conn(ctx); err != nil {
+		return nil, fmt.Errorf("creating snapshot connection: %v", err)
+	}
+
+	// Use context.Background() because we want the Tx to be long lived, we explicitly close it in the close method
+	if s.tx, err = s.snapshotConn.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSnapshot}); err != nil {
+		return nil, fmt.Errorf("starting snapshot transaction: %v", err)
+	}
+
+	var toLSN LSN
+	// capture max LSN _after_ beginning snapshot transaction
+	if err := s.snapshotConn.QueryRowContext(ctx, "SELECT sys.fn_cdc_get_max_lsn()").Scan(&toLSN); err != nil {
+		return nil, err
+	} else if len(toLSN) == 0 {
+		// rare, but possible if the user enabled CDC on a table seconds before running snapshot or the agent has stopped working for some reason
+		return nil, errors.New("unable to captue max_lsn, this can be due to reasons such as the log scanning agent has stopped")
+	}
+
+	return toLSN, nil
+}
+
+// Read starts the process of iterating through each table, reading rows based on maxBatchSize, sending the row as a
+// replication.MessageEvent to the configured publisher.
+func (s *Snapshot) Read(ctx context.Context, maxBatchSize int) error {
+	for _, table := range s.Tables {
+		tablePks, err := s.getTablePrimaryKeys(ctx, table)
+		if err != nil {
+			return err
+		}
+		s.log.Tracef("Primary keys for table '%s': %v", table, tablePks)
+		lastSeenPksValues := map[string]any{}
+		for _, pk := range tablePks {
+			lastSeenPksValues[pk] = nil
+		}
+
+		var numRowsProcessed int
+
+		s.log.Infof("Beginning snapshot process for table '%s'", table.FullName())
+		for {
+			var batchRows *sql.Rows
+			if numRowsProcessed == 0 {
+				batchRows, err = s.querySnapshotTable(ctx, table, tablePks, nil, maxBatchSize)
+			} else {
+				batchRows, err = s.querySnapshotTable(ctx, table, tablePks, &lastSeenPksValues, maxBatchSize)
+			}
+			if err != nil {
+				return fmt.Errorf("failed to execute snapshot table query: %s", err)
+			}
+
+			types, err := batchRows.ColumnTypes()
+			if err != nil {
+				return fmt.Errorf("failed to fetch column types: %w", err)
+			}
+
+			values, mappers := prepSnapshotScannerAndMappers(types)
```

Comment:
Is it really scanner that is returned?

Reply:
The items in the `values` slice all implement the `sql.Scan` interface which is the reason for the `Scanner` part of the function name.